src/vmm/src/devices/virtio/net/device.rs

```diff
@@ -312,25 +363,381 @@ impl Net {
+                            // Queue is empty now, clear readiness and leave.
+                            self.readiness.clear_rx_ready();
+                            break 'read;
+                        };
+
+                        poped_any = true;
+                        // Copy mmds to descriptor chain and write.
+                        let head_index = head.index;
+                        let result = Self::write_to_descriptor_chain(
+                            mem,
+                            &self.rx_mmds_buf[..rx_frame_buf_size],
+                            head,
+                            &self.metrics,
+                        );
+
+                        // Mark the descriptor chain as used. If an error occurred, skip the
+                        // descriptor chain.
+                        let used_len = if result.is_err() {
+                            self.metrics.rx_fails.inc();
+                            0
+                        } else {
+                            self.rx_mmds_read = None;
+                            u32::try_from(rx_frame_buf_size).unwrap()
+                        };
+                        rx_queue
+                            .add_used(mem, head_index, used_len)
+                            .map_err(DeviceError::QueueError)?;
+                        notify_guest = true;
+                        if result.is_ok() {
+                            continue 'read;
+                        }
+                    }
+                    // Give up sending as enough retry.
+                    self.rx_mmds_read = None;
+                }
+            }
+        }
+
+        // Read from tap.
+        if self.readiness.rx_tap_pre_check() {
+            let mem = self.device_state.mem().unwrap();
+            loop {
+                // Check rate limiter.
+                if self.rx_rate_limiter.is_blocked() {
+                    self.metrics.rx_rate_limiter_throttled.inc();
+                    break;
+                }
+
+                // Pop a descriptor chain from the rx queue.
+                tried_pop = true;
+                let Some(head) = rx_queue.pop_or_enable_notification(mem) else {
+                    // Queue is empty now, clear readiness and leave.
+                    self.readiness.clear_rx_ready();
+                    break;
+                };
+                self.rx_rate_limiter.consume(1, TokenType::Ops);
+
+                poped_any = true;
+                // Read from tap to descriptor chain directly.
+                // First write the descriptor chain to IoVecBufferMut.
+                let head_index = head.index;
+
+                macro_rules! push_desc {
+                    ($n:expr) => {
+                        rx_queue
+                            .add_used(mem, head_index, $n)
+                            .map_err(DeviceError::QueueError)?;
+                        notify_guest = true;
+                    };
+                }
```

Comment:
I don't think we really need a macros for this. Also if you address comment above, I think this will go away naturally.